deckhouse-controller/pkg/controller/deckhouse-release/controller.go

```diff
@@ -0,0 +1,455 @@
+	if deckhouseUpdater.ReleasesCount() == 0 {
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	// predict next patch for Deploy
+	deckhouseUpdater.PredictNextRelease()
+
+	// has already Deployed the latest release
+	if deckhouseUpdater.LastReleaseDeployed() {
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	// some release is forced, burn everything, apply this patch!
+	if deckhouseUpdater.HasForceRelease() {
+		deckhouseUpdater.ApplyForcedRelease()
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	if deckhouseUpdater.PredictedReleaseIsPatch() {
+		// patch release does not respect update windows or ManualMode
+		deckhouseUpdater.ApplyPredictedRelease(nil)
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	var windows update.Windows
+	if !deckhouseUpdater.InManualMode() {
+		windows = r.updatePolicy.Get().Update.Windows
+	}
+
+	deckhouseUpdater.ApplyPredictedRelease(windows)
+	return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+}
+
+func (r *deckhouseReleaseReconciler) getDeckhousePods(ctx context.Context) ([]corev1.Pod, error) {
+	var pods corev1.PodList
+	err := r.client.List(
+		ctx,
+		&pods,
+		client.InNamespace("d8-system"),
+		client.MatchingLabels{"app": "deckhouse"},
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("list deckhouse pods: %w", err)
+	}
+
+	filtered := make([]corev1.Pod, 0)
+	for _, pod := range pods.Items {
+		if pod.Status.Phase == corev1.PodFailed {
+			continue
+		}
+
+		filtered = append(filtered, pod)
+	}
+
+	var image, imageID string
+
+	for _, pod := range filtered {
+		// init image and imageID for comparison images/imageIDs across all pods if there are more than one pod in the snapshot
+		if len(image)+len(imageID) == 0 &&
+			len(pod.Spec.Containers[0].Image) != 0 &&
+			len(pod.Status.ContainerStatuses[0].ImageID) != 0 {
+			image, imageID = pod.Spec.Containers[0].Image, pod.Status.ContainerStatuses[0].ImageID
+			continue
+		}
+
+		if image != pod.Spec.Containers[0].Image || imageID != pod.Status.ContainerStatuses[0].ImageID {
+			return nil, fmt.Errorf("deckhouse pods run different images")
+		}
+	}
+
+	return filtered, nil
+}
+
+func (r *deckhouseReleaseReconciler) getDeckhouseDiscoveryData(ctx context.Context) (updater.DeckhouseDiscoveryData, error) {
+	var secret corev1.Secret
+	key := types.NamespacedName{Namespace: "d8-system", Name: "deckhouse-discovery"}
+	err := r.client.Get(ctx, key, &secret)
+	if err != nil {
+		return updater.DeckhouseDiscoveryData{}, fmt.Errorf("get secret %s: %w", key, err)
+	}
+
+	var data = updater.DeckhouseDiscoveryData{
+		ReleaseChannel:         string(secret.Data["releaseChannel"]),
+		ClusterBootstrapping:   true,
+		DisruptionApprovalMode: "Auto",
+		NotificationConfig:     new(updater.NotificationConfig),
+	}
+
+	// default case in template
+	if data.ReleaseChannel == "Unknown" {
+		data.ReleaseChannel = ""
+	}
+
+	clusterBootstrapped, ok := secret.Data["clusterIsBootstrapped"]
```

Comment:
NIT: Don't you want to unmarshal all needed secret data to some struct, and not to ready it directly?
You have to bother about some details, like quotes

Reply:
I haven't found a better way to do this yet